### ISSUE TITLE
Crashes browser EVERYTIME

### ISSUE THREAD

**Issue by mikeerickson** [gligoran/save-to-omnifocus#1]
Every single time I try to use the extension, it crashes browser.

**Comment by gligoran** [comment]
@mikeerickson can I just check that you are actually using THIS extension? There are two with the same name and I had the same problem with the other one.

Mine (by Goran Gligorin): https://chrome.google.com/webstore/detail/save-to-omnifocus/gkjcjolgbgobjdmnocfmojklgpapdnff
The one that crashed my browser (by David Xie): https://chrome.google.com/webstore/detail/save-to-omnifocus/cjbckbkkecemddpbcmcpjcdhgkchacad

**Comment by mikeerickson** [comment]
Yes, I have confirmed it is your extension.  I just removed the extension and tried again and same results.

![Chrome Web Store - save to omnifocus 2020-07-28 10-17-44](https://user-images.githubusercontent.com/183153/88699173-da690980-d0bb-11ea-8c12-f553296bbeb5.png)

**Comment by mikeerickson** [comment]
I tried to do some debugging (not easy) any console I had to script causes the plugin to be invalid (thinking some checksum being done by browser)

**Comment by mikeerickson** [comment]
I have not done any extension work, but know JavaScript so if there are any tips you can offer as it relate to debugging extensions, I am willing to help

**Comment by gligoran** [comment]
This is my first extension as well and thank you very much for offering to help.

Can you provide some more info on what you were doing before it crashed? Did you use the right-click menu to save a page or to save a link? Or did you use the extension button in the top right?
Can you also provide the version of you browser and OS? I'm on the latest Chrome on MacOS Catalina and it works fine for me.
Oh, and also the version of OmniFocus might come in handy.

As far as debugging goes, first try enabling developer mode in the top right corner of chrome://extensions. Then go to the card for this extension and click on the link under the ID (the text might be slightly different):
![image](https://user-images.githubusercontent.com/444829/88774582-e91af380-d183-11ea-9041-c7f7f0278de5.png).

If that doesn't work you can try cloning this repo and using the "Load unpacked" button in top left of chrome://extensions to load in the folder where you cloned the repo to.

**Comment by mikeerickson** [comment]
Can you provide some more info on what you were doing before it crashed? Did you use the right-click menu to save a page or to save a link? Or did you use the extension button in the top right?

> I am using the icon in top right.  Upon clicking button, it just immediately crashes.  


> Can you also provide the version of you browser and OS? I'm on the latest Chrome on MacOS Catalina and it works fine for > me. Oh, and also the version of OmniFocus might come in handy.

Here are all the various versions

- macOS Catalina 10.14.5
- Chrome Version 84.0.4147.89 (Official Build) (64-bit)
- OmniFocus 3.9.1 (v144.4.0)



> As far as debugging goes, first try enabling developer mode in the top right corner of chrome://extensions. 
> Then go to the card for this extension and click on the link under the ID (the text might be slightly different):

Off to do some debugging, hopefully I can get it running and debug

**Comment by mikeerickson** [comment]
@gligoran Well, I have debugged as best as I can, but it still crashes.  The crash occurs in this command, basically the last piece of the extension.

```js
document.body.insertAdjacentHTML('afterend', `<iframe src="${url}" style="display:none" />`);
```

I have verified that body is valid (cant really be invalid) but it still crashes

The following returns **true** so we can safely assume the method _should_ work

```js
document.body.hasOwnProperty('insertAdjacentHTML');  // <-- true
```

Further, I decided to try this with Microsoft Edge (which supports Chrome Extensions) and it crashed as well (which doesn't surprise me seeing as it uses Chromium under the hood)

I wish I could provide more assistance, but I am stumped at this point.

I have moved to use [PopClip](https://pilotmoon.com/popclip/) which has a functional OmniFocus plugin so I have a solution, but i would really like to use the Chrome extension.

One more piece, [Firefox has an extension](https://addons.mozilla.org/en-US/firefox/addon/add-to-omnifocus/) which works fine. Their extension is quite a bit more code, not sure what portion of it is related to Firefox extension boilerplate.

Let me know if you have any further questions, or you are able to get it working.

Regards,

Mike

**Comment by gligoran** [comment]
@mikeerickson I created a PR that attempts to fix this issue: https://github.com/gligoran/save-to-omnifocus/pull/2.

Could you pull the branch `fix-chrome-crash` and load the extension from the folder?

**Comment by mikeerickson** [comment]
OK, I have tried the new build and while it sort of works, when I click the icon, it opens a new chrome tab with the following

<img width="813" alt="Open OmniFocus 2020-08-04 13-11-52" src="https://user-images.githubusercontent.com/183153/89340331-6d71e880-d654-11ea-99f8-320c2ccd0349.png">

**Comment by gligoran** [comment]
[That's a known issue.](https://superuser.com/questions/1492714/chrome-prompting-for-custom-protocol-handlers-every-time-after-update) There used to be a checkbox on that popup that allowed you to remember your choice, but it was removed with Chrome 77 as there was no way to change it.

I found a way to make that popup go away, by adding the `omnifocus://` protocol to the `[URLWhitelist](https://cloud.google.com/docs/chrome-enterprise/policies/?policy=URLWhitelist) via this command:

```
defaults write com.google.Chrome URLWhitelist -array-add 'omnifocus://*'
```

Once added you can see it in `chrome://policy/`.

**Comment by mikeerickson** [comment]
@gligoran We are getting closer!

After adding `omnifocus://` to the chrome policy, I am now able to click the button and the OmniFocus Quick Entry is displayed with the subject / title working.

However, the notes field does not have the URL. I took a look at the source code and fixed the issue by replacing the use of `&amp;` with the actual `ampersand (%)` to build the URL

In the `saveToOmniFocus` method, take a look at the following

```js
const url = task.name
    ? `omnifocus:///add?name=${encodeURIComponent(task.name)}&note=${encodeURIComponent(task.note)}`
    : `omnifocus:///add?name=${encodeURIComponent(task.note)}`;
```

Using this change locally, everything works as expected!

**Comment by gligoran** [comment]
@mikeerickson The fixed version (1.0.2) should be out now.
